Fixed #23813 -- Added checks for common URL pattern errors

alasdairnicol opened the pull request.

jarshwah commented on django/core/checks/urls.py:
```diff
@@ -0,0 +1,51 @@
+from . import Tags, Warning, register
+
+W001 = Warning(
+    "In your url patterns, you have used include with a regex containing a $. "
```
I'm not too keen on beginning each warning with "In your url patterns, ..".

How about "Your url patterns .." ?

"Your url patterns have used `include` with a regex containing a '$'. " .. 

"Your url patterns have a regex beginning with a '/'." ..

"Your url patterns have a pattern with a name containing a ':'." ..

alasdairnicol replied:
Good idea, I've simplified the text. I went for "One of your url patterns..." because it's simpler than "Your url patterns have a patten...".

timgraham commented on django/core/checks/urls.py:
```diff
@@ -0,0 +1,51 @@
+# -*- coding: utf-8 -*-
```
We usually only include this if there are non-ASCII characters in the file.

timgraham commented on django/core/checks/urls.py:
```diff
@@ -0,0 +1,51 @@
+from . import Tags, Warning, register
+
+W001 = Warning(
+    "One of your url patterns uses include with a regex containing a '$'. "
```
URL patterns

timgraham commented on docs/ref/checks.txt:
```diff
@@ -81,6 +81,7 @@ Django's system checks are organized using the following tags:
 * ``security``: Checks security related configuration.
 * ``templates``: Checks template related configuration.
 * ``caches``: Checks cache related configuration.
+* ``urls``: Checks url configuration
```
URL

timgraham commented:
The main problem I see here is that the error doesn't point to the offending pattern. In a large project, it could be very tedious to track down the problem.

alasdairnicol commented:
Thanks for the review. I'll fix up the minor changes tomorrow. It should be straight forward to report the regex that is causing the errors, but I'm not sure it will be as easy to report which module the errors are in. Hopefully that will be more obvious when I look at the code tomorrow.

jarshwah commented on django/core/checks/urls.py:
```diff
@@ -0,0 +1,87 @@
+    regex_pattern = pattern.regex.pattern
+    if regex_pattern.endswith('$') and not regex_pattern.endswith('\$'):
+        warning = Warning(
+            "Your URL pattern {} uses include with a regex ending with a '$'. "
```
I think the formatted pattern should be wrapped in quotes. "Your URL pattern '{}' uses..".

jarshwah commented on django/core/checks/urls.py:
```diff
@@ -0,0 +1,87 @@
+    regex_pattern = pattern.regex.pattern
+    if regex_pattern.startswith('/') or regex_pattern.startswith('^/'):
+        warning = Warning(
+            "Your URL pattern {} has a regex beginning with a '/'. "
```
As above, wrap the format in quotes. "Your URL pattern '{}' has .."

jarshwah commented on django/core/checks/urls.py:
```diff
@@ -0,0 +1,87 @@
+    """
+    if pattern.name is not None and ":" in pattern.name:
+        warning = Warning(
+            "Your URL pattern {} has a name including a ':'. Remove the colon, to "
```
As above, wrap the format in quotes. "Your URL pattern '{}' has .."

jarshwah commented:
I've made 3 suggestions for wrapping the format pattern in quotes. Feel free to add yourself to AUTHORS too if you like. Squash/Rebase, and hopefully we can squeeze this in for 1.9 which alpha's today/tomorrow.

alasdairnicol commented:
@jarshwah I've wrapped the regex in quotes - I did it slightly differently than you suggested, so that it displays

    '^$' [name='name']

instead of 

    '^$ [name='name']'

I have squashed and rebased. No need to add myself to AUTHORS, I'm already there :-)

jarshwah commented:
Merged in fe3fc52. Thanks!

timgraham commented on django/core/checks/urls.py:
```diff
@@ -0,0 +1,87 @@
+
+def describe_pattern(pattern):
+    """
+    Formats the URL pattern for display in warning messages
```
For future reference, we are using PEP 257 style verbs for new docstrings "Format...", etc.. Also, it's nice to be consistent with trailing punctuation.

alasdairnicol replied:
I vaguely remembered a discussion about doc strings, so I checked the style guide, but still got it wrong because the advice changed between 1.8 and dev!  Sorry about the punctuation.

timgraham commented:
I guess it might be nice to move the URLconfs into a module to avoid cluttering up the main directory as more checks are added.

alasdairnicol commented:
I can fix up the docstrings and move the URL confs to a directory this
evening (UK) Does dropping the url_ prefix sound ok, i.e.

urls_include.py -> urls/include.py
urls_name.py -> urls/name.py
...

On 21 September 2015 at 15:05, Tim Graham <notifications@github.com> wrote:

> I guess it might be nice to move the URLconfs into a module to avoid
> cluttering up the main directory as more checks are added.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/django/django/pull/5301#issuecomment-141991761>.
>

timgraham commented:
Yep, thanks.

alasdairnicol commented:
> urls_include.py -> urls/include.py

Actually, the directory can't be in a directory 'urls' since the checks are
already in urls.py. So could do e.g.

urlconfs/include.py

On 21 September 2015 at 15:14, Alasdair Nicol <alasdair@thenicols.net>
wrote:

> I can fix up the docstrings and move the URL confs to a directory this
> evening (UK) Does dropping the url_ prefix sound ok, i.e.
>
> urls_include.py -> urls/include.py
> urls_name.py -> urls/name.py
> ...
>
> On 21 September 2015 at 15:05, Tim Graham <notifications@github.com>
> wrote:
>
>> I guess it might be nice to move the URLconfs into a module to avoid
>> cluttering up the main directory as more checks are added.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/django/django/pull/5301#issuecomment-141991761>.
>>
>
>
